flexget/components/ftp/sftp_client.py

```diff
@@ -0,0 +1,423 @@
+from typing import Callable, List, Optional
+import logging
+import os
+import posixpath
+import time
+from functools import partial
+from urllib.parse import quote, urljoin
+
+from flexget import plugin
+from flexget.entry import Entry
+
+# retry configuration constants
+CONNECT_TRIES: int = 3
+RETRY_INTERVAL: int = 15
+RETRY_STEP: int = 5
+SOCKET_TIMEOUT: int = 15
+
+try:
+    import pysftp
+    logging.getLogger("paramiko").setLevel(logging.ERROR)
+except ImportError:
+    pysftp = None
+
+local_path = os.path
+remote_path = posixpath  # pysftp uses POSIX style paths
+
+
+class SftpClient:
+    def __init__(self, logger: logging.Logger, host: str, port: int, username: str, password: Optional[str] = None,
+                 private_key: Optional[str] = None, private_key_pass: Optional[str] = None):
+
+        if not pysftp:
+            raise plugin.DependencyError(
+                issued_by='sftp_client',
+                missing='pysftp',
+                message='sftp client requires the pysftp Python module.',
+            )
+
+        self._logger: logging.Logger = logger
```

Comment:
I could see it being useful if you used this module in another plugin or component, since it means it'll inherit whatever context the calling module gave it. Not sure if that's considered desirable or not.